File: python/40_self_describing_numbers.py
import re
import sys

def is_self_describing(n):
    for i in range(len(n)):
        c = n[i]

        if int(c) != len(re.findall(str(i), n)):
            return False

    return True

with open(sys.argv[1], 'r') as fh:
    for line in fh.readlines():
        line = line.strip()
        if line == '':
            continue

        print 1 if is_self_describing(line) else 0

File: perl/40_self_describing_numbers.pl
#!/usr/bin/env perl

use strict;
use warnings;

open my $fh, '<', $ARGV[0];

sub is_self_describing {
	my $n = shift @_;

	for ( my $i = 0; $i != length $n; $i++ ) {
		my $c = substr $n, $i, 1;

		my @m = $n =~ m/$i/g;
		return 0 if $c != scalar @m;
	}

	return 1;
}

for my $line ( <$fh> ) {
	chomp $line;

	next if $line eq '';

	print is_self_describing( $line ), "\n";
}

close $fh;

exit 0;

